ts/sagas/api.ts

```diff
@@ -0,0 +1,37 @@
+import { Either, left, right } from "fp-ts/lib/Either";
+import { Effect } from "redux-saga";
+import { call, put } from "redux-saga/effects";
+
+import { BasicResponseTypeWith401 } from "../api/backend";
+import { sessionExpired } from "../store/actions/authentication";
+
+/**
+ * A generator that wrap an API Call and handle the response status.
+ * If the response status is 401 the user must be logged out.
+ */
+export function* callApi<
+  P,
+  RT,
+  T extends Promise<BasicResponseTypeWith401<RT> | undefined>
+>(
+  apiCall: (param: P) => T,
+  param: P
+): IterableIterator<Effect | Either<Error, RT>> {
+  const response: BasicResponseTypeWith401<RT> | undefined = yield call(
+    apiCall,
+    param
+  );
+
+  // Not able to decode the response
+  if (!response) {
+    return left(Error());
```

Comment:
it would be safer for this wrapper to avoid changing the semantic of the response it's not interested about, so if the underlying client returns `undefined` in case is unable to decode the response, this wrapper should return the same value

Reply:
What you think about just catching the 401 avoiding to change the response at all, like this

```
if (response && response.status === 401) {
    yield put(sessionExpired());
  }
  return response;
```

Reply:
that would be ideal :)

then you can name this method something like `withHandleLogout(...originalAPIcall...)`